In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression, Ridge, Lasso
from sklearn.metrics import mean_squared_error, r2_score
import numpy as np

In [2]:
from google.colab import files
uploaded = files.upload()


Saving Hitters.csv to Hitters.csv


In [3]:
df = pd.read_csv('Hitters.csv')
df.head()

AtBat  Hits  HmRun  Runs  RBI  Walks  Years  CAtBat  CHits  CHmRun  CRuns  \
0    293    66      1    30   29     14      1     293     66       1     30   
1    315    81      7    24   38     39     14    3449    835      69    321   
2    479   130     18    66   72     76      3    1624    457      63    224   
3    496   141     20    65   78     37     11    5628   1575     225    828   
4    321    87     10    39   42     30      2     396    101      12     48   

   CRBI  CWalks League Division  PutOuts  Assists  Errors  Salary NewLeague  
0    29      14      A        E      446       33      20     NaN         A  
1   414     375      N        W      632       43      10   475.0         N  
2   266     263      A        W      880       82      14   480.0         A  
3   838     354      N        E      200       11       3   500.0         N  
4    46      33      N        E      805       40       4    91.5         N

In [4]:
df.info()
df.describe()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 322 entries, 0 to 321
Data columns (total 20 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   AtBat      322 non-null    int64  
 1   Hits       322 non-null    int64  
 2   HmRun      322 non-null    int64  
 3   Runs       322 non-null    int64  
 4   RBI        322 non-null    int64  
 5   Walks      322 non-null    int64  
 6   Years      322 non-null    int64  
 7   CAtBat     322 non-null    int64  
 8   CHits      322 non-null    int64  
 9   CHmRun     322 non-null    int64  
 10  CRuns      322 non-null    int64  
 11  CRBI       322 non-null    int64  
 12  CWalks     322 non-null    int64  
 13  League     322 non-null    object 
 14  Division   322 non-null    object 
 15  PutOuts    322 non-null    int64  
 16  Assists    322 non-null    int64  
 17  Errors     322 non-null    int64  
 18  Salary     263 non-null    float64
 19  NewLeague  322 non-null    object 
dtypes: float64

AtBat        Hits       HmRun        Runs         RBI       Walks  \
count  322.000000  322.000000  322.000000  322.000000  322.000000  322.000000   
mean   380.928571  101.024845   10.770186   50.909938   48.027950   38.742236   
std    153.404981   46.454741    8.709037   26.024095   26.166895   21.639327   
min     16.000000    1.000000    0.000000    0.000000    0.000000    0.000000   
25%    255.250000   64.000000    4.000000   30.250000   28.000000   22.000000   
50%    379.500000   96.000000    8.000000   48.000000   44.000000   35.000000   
75%    512.000000  137.000000   16.000000   69.000000   64.750000   53.000000   
max    687.000000  238.000000   40.000000  130.000000  121.000000  105.000000   

            Years       CAtBat        CHits      CHmRun        CRuns  \
count  322.000000    322.00000   322.000000  322.000000   322.000000   
mean     7.444099   2648.68323   717.571429   69.490683   358.795031   
std      4.926087   2324.20587   654.472627   86.266061   334.105886   
min      1.000000     19.00000     4.000000    0.000000     1.000000   
25%      4.000000    816.75000   209.000000   14.000000   100.250000   
50%      6.000000   1928.00000   508.000000   37.500000   247.000000   
75%     11.000000   3924.25000  1059.250000   90.000000   526.250000   
max     24.000000  14053.00000  4256.000000  548.000000  2165.000000   

              CRBI       CWalks      PutOuts     Assists      Errors  \
count   322.000000   322.000000   322.000000  322.000000  322.000000   
mean    330.118012   260.239130   288.937888  106.913043    8.040373   
std     333.219617   267.058085   280.704614  136.854876    6.368359   
min       0.000000     0.000000     0.000000    0.000000    0.000000   
25%      88.750000    67.250000   109.250000    7.000000    3.000000   
50%     220.500000   170.500000   212.000000   39.500000    6.000000   
75%     426.250000   339.250000   325.000000  166.000000   11.000000   
max    1659.000000  1566.000000  1378.000000  492.000000   32.000000   

            Salary  
count   263.000000  
mean    535.925882  
std     451.118681  
min      67.500000  
25%     190.000000  
50%     425.000000  
75%     750.000000  
max    2460.000000

In [5]:
df.isnull().sum()


AtBat         0
Hits          0
HmRun         0
Runs          0
RBI           0
Walks         0
Years         0
CAtBat        0
CHits         0
CHmRun        0
CRuns         0
CRBI          0
CWalks        0
League        0
Division      0
PutOuts       0
Assists       0
Errors        0
Salary       59
NewLeague     0
dtype: int64

In [6]:
df = df.dropna()


In [7]:
df.isnull().sum()


AtBat        0
Hits         0
HmRun        0
Runs         0
RBI          0
Walks        0
Years        0
CAtBat       0
CHits        0
CHmRun       0
CRuns        0
CRBI         0
CWalks       0
League       0
Division     0
PutOuts      0
Assists      0
Errors       0
Salary       0
NewLeague    0
dtype: int64

In [8]:
df = pd.get_dummies(df, drop_first=True)


In [9]:
df.head()
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 263 entries, 1 to 321
Data columns (total 20 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   AtBat        263 non-null    int64  
 1   Hits         263 non-null    int64  
 2   HmRun        263 non-null    int64  
 3   Runs         263 non-null    int64  
 4   RBI          263 non-null    int64  
 5   Walks        263 non-null    int64  
 6   Years        263 non-null    int64  
 7   CAtBat       263 non-null    int64  
 8   CHits        263 non-null    int64  
 9   CHmRun       263 non-null    int64  
 10  CRuns        263 non-null    int64  
 11  CRBI         263 non-null    int64  
 12  CWalks       263 non-null    int64  
 13  PutOuts      263 non-null    int64  
 14  Assists      263 non-null    int64  
 15  Errors       263 non-null    int64  
 16  Salary       263 non-null    float64
 17  League_N     263 non-null    bool   
 18  Division_W   263 non-null    bool   
 19  NewLeague_N  

In [10]:
df[["League_N", "Division_W", "NewLeague_N"]] = df[["League_N", "Division_W", "NewLeague_N"]].astype(int)


In [11]:
X = df.drop("Salary", axis=1)
y = df["Salary"]

In [12]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [13]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [14]:
lr_model = LinearRegression()
lr_model.fit(X_train, y_train)

LinearRegression()

In [16]:
y_pred = lr_model.predict(X_test)

In [18]:
from sklearn.metrics import mean_squared_error, r2_score


In [21]:
mse = mean_squared_error(y_test, y_pred)
rmse = mse ** 0.5
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error :" , mse )
print("Root Mean Squared Error :" ,  rmse )
print("R-squared :" , r2 )

Mean Squared Error : 128284.34549672325
Root Mean Squared Error : 358.1680408645127
R-squared : 0.29074518557981566


In [26]:
from sklearn.linear_model import Ridge
from sklearn.metrics import mean_squared_error, r2_score

# Modeli oluştur
ridge_model = Ridge(alpha=1.0)

# Eğitim verisi ölçeklenmiş olmalı!
ridge_model.fit(X_train_scaled, y_train)

Ridge()

In [34]:
from sklearn.linear_model import RidgeCV

# Farklı alpha değerlerini deneyelim
ridgecv = RidgeCV(alphas=[0.01, 0.1, 1.0, 10.0, 100.0])
ridgecv.fit(X_train_scaled, y_train)

# En iyi alpha değerini bulalım
print("En iyi alpha:", ridgecv.alpha_)

# Test verisiyle tahmin yapalım
y_pred_ridge = ridgecv.predict(X_test_scaled)

# Performans ölçümleri
mse_ridge = mean_squared_error(y_test, y_pred_ridge)
rmse_ridge = mse_ridge ** 0.5
r2_ridge = r2_score(y_test, y_pred_ridge)

print("🔵 Ridge Regression (RidgeCV ile alpha optimize edilmiş)")
print("MSE:", mse_ridge)
print("RMSE:", rmse_ridge)
print("R^2:", r2_ridge)

En iyi alpha: 100.0
🔵 Ridge Regression (RidgeCV ile alpha optimize edilmiş)
MSE: 146501.31952267422
RMSE: 382.75490790148496
R^2: 0.1900277014468572


In [31]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

In [33]:
lasso_model = Lasso(alpha=1.0)
lasso_model.fit(X_train_scaled, y_train)
y_pred_lasso = lasso_model.predict(X_test_scaled)

mse_lasso = mean_squared_error(y_test, y_pred_lasso)
rmse_lasso = mse_lasso ** 0.5
r2_lasso = r2_score(y_test, y_pred_lasso)

print("\n🟣 Lasso Regression")
print("MSE:", mse_lasso)
print("RMSE:", rmse_lasso)
print("R^2:", r2_lasso)


🟣 Lasso Regression
MSE: 126527.01699679394
RMSE: 355.70636344714717
R^2: 0.30046105304802906


In [39]:
# Kategorik sütunları one-hot encoding ile dönüştür
df_encoded = pd.get_dummies(df, columns=['League_N', 'Division_W', 'NewLeague_N'], drop_first=True)

# Özellikleri ve hedefi ayıralım
X = df_encoded.drop(columns=['Salary'])
y = df_encoded['Salary']

In [40]:
df_encoded['AtBat_Hits_Ratio'] = df_encoded['AtBat'] / (df_encoded['Hits'] + 1)

In [41]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Veriyi ölçekleyelim
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# Ridge modelini yeniden eğitelim
ridge_model = Ridge(alpha=0.1)
ridge_model.fit(X_train_scaled, y_train)

# Tahmin yapalım
y_pred_ridge = ridge_model.predict(X_test_scaled)

# Performans ölçümleri
mse_ridge = mean_squared_error(y_test, y_pred_ridge)
rmse_ridge = mse_ridge ** 0.5
r2_ridge = r2_score(y_test, y_pred_ridge)

print("🔵 Ridge Regression (Özellik Mühendisliği Sonrası)")
print("MSE:", mse_ridge)
print("RMSE:", rmse_ridge)
print("R^2:", r2_ridge)

🔵 Ridge Regression (Özellik Mühendisliği Sonrası)
MSE: 127645.1141074537
RMSE: 357.27456403647557
R^2: 0.2942793497727453


In [43]:
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)

X_train_poly, X_test_poly, y_train, y_test = train_test_split(X_poly, y, test_size=0.2, random_state=42)

In [44]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train_poly)
X_test_scaled = scaler.transform(X_test_poly)

In [45]:
ridge_model_poly = Ridge(alpha=100.0)

# Modeli eğit
ridge_model_poly.fit(X_train_scaled, y_train)

# Tahmin yap
y_pred_poly = ridge_model_poly.predict(X_test_scaled)

In [47]:
mse_poly = mean_squared_error(y_test, y_pred_poly)
rmse_poly = mse_poly ** 0.5
r2_poly = r2_score(y_test, y_pred_poly)

# Sonuçları yazdır
print(" Ridge Regression (Polinom Özellikler ile Sonuç)")
print("MSE:", mse_poly)
print("RMSE:", rmse_poly)
print("R^2:", r2_poly)

 Ridge Regression (Polinom Özellikler ile Sonuç)
MSE: 105687.42995378877
RMSE: 325.0960318948676
R^2: 0.41567836489967225


Polinom Özelliklerle Ridge Regression:
Son olarak, modelin öğrenebileceği karmaşık ilişkileri artırmak için polinom özellikler ekledik (degree=2):

Model artık sadece düz çizgiler değil, eğriler gibi daha karmaşık desenleri de öğrenebildi.

Bu sayede:

 R² ≈ 0.416 → Model verinin %41.6'sını açıklayabiliyor.
 MSE ve RMSE ciddi şekilde düştü.
 Modelin genel doğruluğu bariz şekilde iyileşti.

Modeli doğru kurmak sadece algoritmayı çağırmak değilmiş:

Doğru model + doğru alpha + doğru özellikler = Güçlü tahmin gücü